File: tools/AFiDTools.jl
module AFiDTools

export
    read_grid,
    read_mean,
    read_cut,
    continua_master_from_input,
    write_continua,
    tmean,
    xmean,
    ddx,
    coarsen,
    refine,
    xmf_cuts,
    xmf_fields

using
    DelimitedFiles,
    HDF5,
    Interpolations,
    Statistics,
    LightXML,
    Printf

"""
    function read_grid(folder::String)
    
Returns an named tuple containing staggered coordinates `xm`, `ym`, `zm`,
cell-edge coordinates e.g. `xc`, and refined grid coordinates e.g. `xmr`
"""
function read_grid(folder::String)
    filename = folder*"/outputdir/cordin_info.h5"
    fileid = h5open(filename,"r")
    xm, xc = read(fileid["xm"]), read(fileid["xc"])
    ym, zm = read(fileid["ym"]), read(fileid["zm"])
    if "xmr" in keys(fileid)
        xmr, xcr = read(fileid["xmr"]), read(fileid["xcr"])
        ymr, zmr = read(fileid["ymr"]), read(fileid["zmr"])
    else
        xmr, xcr, ymr, zmr = [], [], [], []
    end
    close(fileid)
    if length(ym)>1
        dy = ym[2] - ym[1]
        yc = 0:dy:dy*length(ym)
    else
        yc = [0,2*ym[1]]
    end
    if length(ymr)>1
        dyr = ymr[2] - ymr[1]
        ycr = 0:dyr:dyr*length(ymr)
    elseif length(ymr)==1
        ycr = [0,2*ymr[1]]
    else
        ycr = []
    end
    if length(zm)>1
        dz = zm[2] - zm[1]
        zc = 0:dz:dz*length(zm)
    else
        zc = [0,2*zm[1]]
    end
    if length(zmr)>1
        dzr = zmr[2] - zmr[1]
        zcr = 0:dzr:dzr*length(zmr)
    elseif length(zmr)==1
        zcr = [0,2*zmr[1]]
    else
        zcr = []
    end
    return (xm=xm, xmr=xmr, xc=xc, xcr=xcr, ym=ym, ymr=ymr, yc=yc, ycr=ycr, zm=zm, zmr=zmr, zc=zc, zcr=zcr)
end

"""
    function read_mean(folder::String,varname::String)
    
Returns a 2D array containing a time series of mean profiles f̄(x,t) of `varname`
from the simulation output in `folder`.
"""
function read_mean(folder::String,varname::String)
    filename = folder*"/outputdir/means.h5"
    fileid = h5open(filename,"r")
    varlist = keys(fileid)
    if varname in varlist
        list = keys(fileid[varname])
        Nsamp = size(list)[1]
        nxx = length(fileid[varname*"/"*list[1]])
        var = zeros((nxx,Nsamp))
        i = 1
        for num in list
            var[:,i] = read(fileid[varname*"/"*num])
            i = i + 1;
        end
    else
        varstring = string()
        for var in varlist
            varstring *= var*", "
        end
        error(varname*" not a variable in means.h5. Valid options are "*varstring)
    end
    close(fileid)
    return var
end

"""
    function read_cut(folder::String,varname::String,idx::Int,plane::String)
    
Returns an array slice of `varname` ("vx", "vy", "vz", "temp", "sal", or "phi")
from a `plane` ("x", "y", or "z") at time index `idx` from the simulation
output in `folder`.
"""
function read_cut(folder::String,varname::String,idx::Int,plane::String)
    filename = folder*"/outputdir/flowmov/movie_"*plane*"cut.h5"
    fileid = h5open(filename,"r")
    var = read(fileid[varname*"/"*string(idx,pad=5)])
    close(fileid)
    return var
end

function continua_master_from_input(folder::String)
    input_data = readdlm(folder*"/bou.in")
    nx, ny, nz = input_data[3,1:3] .+ 1
    nxr, nyr, nzr = input_data[6,2:4] .+ 1
    ylen, zlen = input_data[21,2:3]
    istr3, str3, istr3r = input_data[24,1:3]
    time = 0.0
    fname = folder*"/outputdir/continua_master.h5"
    fid = h5open(fname,"w")
    fid["nx"], fid["ny"], fid["nz"] = nx, ny, nz
    fid["nxr"], fid["nyr"], fid["nzr"] = nxr, nyr, nzr
    fid["ylen"], fid["zlen"] = ylen, zlen
    fid["istr3"], fid["str3"], fid["istr3r"] = istr3, str3, istr3r
    fid["time"] = time
    close(fid)
    return
end

function write_continua(folder::String,varname::String,var::Array{Float64, 3})
    fname = folder*"/outputdir/continua_"*varname*".h5"
    fid = h5open(fname,"w")
    fid["var"] = var
    close(fid)
    return
end

"""
    function tmean(var::Array{Float64})
    
Computes the time average of data obtained from the `read_mean()` function.
Passing a 2D array produces a time series `Vector`, whereas passing a 1D
array gives the mean and standard deviation from the time series as a named tuple.
"""
function tmean(var::Matrix{Float64})
    if size(var)[2] < 1350
        A = mean(var[:,150:end], dims=2)
    else
        A = mean(var[:,end-1250:end], dims=2)
    end
    return A
end
function tmean(var::Vector{Float64})
    if length(var) < 1350
        tmp = var[150:end]
    else
        tmp = var[end-1250:end]
    end
    return (mean=mean(tmp), std=std(tmp))
end

"""
    function xmean(var::Array{Float64}, grid)

Computes the domain average of the mean profile time series data
obtained from the `read_mean()` function.
Requires an input `grid` that matches the output from `read_grid()`.
"""
function xmean(var::Array{Float64}, grid)
    if size(var)[1]==length(grid.xm)
        dx = grid.xc[2:end] - grid.xc[1:end-1]
    else
        dx = grid.xcr[2:end] - grid.xcr[1:end-1]
    end
    mvar = sum(dx .* var, dims=1)[:]
    return mvar
end

"""
    function ddx(f::Array{Float64}, x::Vector{Float64}; BClow::Float64=0.0, BChigh::Float64=0.0)
Computes the x-derivative of the mean profile quantities generated by
`read_mean()` when the relevant x-coordinates `x` are passed to the function.
Optional arguments for the lower and upper boundary values can be passed as
`BClow` and `BChigh`. By default, these are set to zero.
"""
function ddx(f::Matrix{Float64}, x::Vector{Float64}; BClow::Float64=0.0, BChigh::Float64=0.0)
    df = zeros(size(f))
    df[1,:] = (f[2,:] .- BClow)./(x[2] - 0)
    df[2:end-1,:] = (f[3:end,:] .- f[1:end-2,:])./(x[3:end] .- x[1:end-2])
    df[end,:] = (BChigh .- f[end-1,:])./(1 - x[end-1])
    return df
end
function ddx(f::Vector{Float64}, x::Vector{Float64}; BClow::Float64=0.0, BChigh::Float64=0.0)
    df = zeros(size(f))
    df[1] = (f[2] - BClow)./(x[2] - 0)
    df[2:end-1] = (f[3:end] - f[1:end-2])./(x[3:end] - x[1:end-2])
    df[end] = (BChigh - f[end-1])./(1 - x[end-1])
    return df
end

"""
    function coarsen(var::Array{Float64}, grid)
Interpolates a `read_mean()` quantity `var` from the base grid `grid.xm`
to the refined grid `grid.xmr`.
"""
function coarsen(var::Matrix{Float64}, grid)
    nodes = (grid.xmr,)
    Ns = size(var)[2]
    varc = zeros(length(grid.xm),Ns)
    for i=1:Ns
        itp = interpolate(nodes, var[:,i], Gridded(Linear()))
        varc[:,i] = itp.(grid.xm)
    end
    return varc
end
function coarsen(var::Vector{Float64}, grid)
    nodes = (grid.xmr,)
    itp = interpolate(nodes, var[:], Gridded(Linear()))
    varc = itp.(grid.xm)
    return varc
end
"""
    function refine(var::Array{Float64}, grid)
Interpolates a `read_mean()` quantity `var` from the refined grid `grid.xmr`
to the base grid `grid.xm`.
"""
function refine(var::Matrix{Float64}, grid)
    x = vcat(0, grid.xm, 1)
    nodes = (x,)
    Ns = size(var)[2]
    varr = zeros(length(grid.xmr),Ns)
    tmpvar = vcat(transpose(zeros(Ns)), var, transpose(zeros(Ns)))
    for i=1:Ns
        itp = interpolate(nodes, tmpvar[:,i], Gridded(Linear()))
        varr[:,i] = itp.(grid.xmr)
    end
    return varr
end
function refine(var::Vector{Float64}, grid)
    x = vcat(0, grid.xm, 1)
    nodes = (x,)
    tmpvar = vcat(0, var[:], 0)
    itp = interpolate(nodes, tmpvar, Gridded(Linear()))
    varr = itp.(grid.xmr)
    return varr
end

"""
    function xmf_cuts(folder::String)
"""
function xmf_cuts(folder::String)

    grid = read_grid(folder)
    nxm, nym, nzm = length(grid.xm), length(grid.ym), length(grid.zm)
    dims = Dict(string(key) => length(grid[key]) for key in keys(grid))
    cutdims = Dict{String,Int64}()
    grids = Dict(
        "vx" => Dict("x"=>"xc", "y"=>"ym", "z"=>"zm"),
        "vy" => Dict("x"=>"xm", "y"=>"ym", "z"=>"zm"),
        "vz" => Dict("x"=>"xm", "y"=>"ym", "z"=>"zm"),
        "temp" => Dict("x"=>"xm", "y"=>"ym", "z"=>"zm"),
        "sal" => Dict("x"=>"xmr", "y"=>"ymr", "z"=>"zmr"),
        "phi" => Dict("x"=>"xmr", "y"=>"ymr", "z"=>"zmr")
    )

    dt = readdlm(folder*"/bou.in")[18,2]
    cut_vals = Dict("x"=>grid.xm[1], "y"=>grid.yc[end]/2, "z"=>grid.zc[end]/2)

    for plane in ["x","y","z"]
        f = h5open(folder*"/outputdir/flowmov/movie_"*plane*"cut.h5","r")
        varlist = keys(f)
        samplist = parse.(Int, keys(f["temp"]))
        close(f)
        for var in varlist
            xdoc = XMLDocument()
            xroot= create_root(xdoc, "Xdmf")
            droot= new_child(xroot, "Domain")
            troot= new_child(droot, "Grid")
            set_attributes(troot; GridType="Collection", CollectionType="Temporal")
            for i in samplist
                xcut = new_child(troot, "Grid")
                set_attributes(xcut)
                time = new_child(xcut, "Time")
                set_attributes(time; Value=@sprintf("%.2f", i*dt))
                for k in keys(grids[var])
                    cutdims[k] = dims[grids[var][k]]
                end
                cutdims[plane] = 1
                mesh = new_child(xcut, "Topology")
                set_attributes(mesh; TopologyType="3DRectMesh", Dimensions=@sprintf("%i %i %i", cutdims["z"], cutdims["y"], cutdims["x"]))
                geom = new_child(xcut, "Geometry")
                set_attributes(geom; GeometryType="VXVYVZ")
                for grid in ["x","y","z"]
                    grid_data = new_child(geom, "DataItem")
                    if plane==grid
                        set_attributes(grid_data; Dimensions="1")
                        add_text(grid_data, @sprintf("%.3f", cut_vals[plane]))
                    else
                        set_attributes(grid_data; Dimensions=@sprintf("%i", cutdims[grid]), Format="HDF")
                        add_text(grid_data, "../cordin_info.h5:/"*grids[var][grid])
                    end
                end
                att = new_child(xcut, "Attribute")
                set_attributes(att; AttributeType="Scalar", Center="Node", Name=var)
                data = new_child(att, "DataItem")
                set_attributes(data; Dimensions=@sprintf("%i %i %i", cutdims["z"], cutdims["y"], cutdims["x"]), Format="HDF")
                add_text(data, "movie_"*plane*"cut.h5:/"*var*"/"*@sprintf("%05i", i))
            end
            save_file(xdoc, folder*"/outputdir/flowmov/"*var*"_"*plane*"cut.xmf")
        end
    end
end

"""
    function xmf_fields(folder::String)
"""
function xmf_fields(folder::String)

    grid = read_grid(folder)
    nxm, nym, nzm = length(grid.xm), length(grid.ym), length(grid.zm)
    nxmr, nymr, nzmr = length(grid.xmr), length(grid.ymr), length(grid.zmr)
    dx, dy, dz = 1/nxm, grid.yc[end]/nym, grid.zc[end]/nzm
    if nxmr!=0
        dxr, dyr, dzr = 1/nxmr, grid.ycr[end]/nymr, grid.zcr[end]/nzmr
    end
    dims = Dict(string(key) => length(grid[key]) for key in keys(grid))
    fielddims = Dict{String,Int64}()
    grids = Dict(
        "vx" => Dict("x"=>"xc", "y"=>"ym", "z"=>"zm"),
        "vy" => Dict("x"=>"xm", "y"=>"ym", "z"=>"zm"),
        "vz" => Dict("x"=>"xm", "y"=>"ym", "z"=>"zm"),
        "temp" => Dict("x"=>"xm", "y"=>"ym", "z"=>"zm"),
        "sal" => Dict("x"=>"xmr", "y"=>"ymr", "z"=>"zmr"),
        "phi" => Dict("x"=>"xmr", "y"=>"ymr", "z"=>"zmr")
    )

    dt = readdlm(folder*"/bou.in")[18,2]
    varlist = unique(replace(t->t[7:end-3],readdir(folder*"/outputdir/fields")))
    for var in varlist
        samplist = parse.(Int, replace(t->t[1:5], filter(t->t[7:end-3]=="temp",readdir(folder*"/outputdir/fields"))))
        sort!(samplist)
        xdoc = XMLDocument()
        xroot= create_root(xdoc, "Xdmf")
        droot= new_child(xroot, "Domain")
        troot= new_child(droot, "Grid")
        set_attributes(troot; GridType="Collection", CollectionType="Temporal")
        for i in samplist
            field = new_child(troot, "Grid")
            set_attributes(field)
            time = new_child(field, "Time")
            set_attributes(time; Value=@sprintf("%.2f", i*dt))
            for k in keys(grids[var])
                fielddims[k] = dims[grids[var][k]]
                if occursin("yc",grids[var][k]) || occursin("zc",grids[var][k])
                    fielddims[k] = fielddims[k] - 1
                end
            end
            mesh = new_child(field, "Topology")
            set_attributes(mesh; TopologyType="3DCoRectMesh", Dimensions=@sprintf("%i %i %i", fielddims["z"], fielddims["y"], fielddims["x"]))
            geom = new_child(field, "Geometry")
            set_attributes(geom; GeometryType="ORIGIN_DXDYDZ")
            origin = new_child(geom, "DataItem")
            set_attributes(origin; Dimensions="3")
            add_text(origin, @sprintf("%.5f %.5f %.5f", grid[Symbol(grids[var]["x"])][1], grid[Symbol(grids[var]["y"])][1], grid[Symbol(grids[var]["z"])][1]))
            step = new_child(geom, "DataItem")
            set_attributes(step; Dimensions="3")
            if occursin(var, "phisal")
                add_text(step, @sprintf("%.5f %.5f %.5f", dxr, dyr, dzr))
            else
                add_text(step, @sprintf("%.5f %.5f %.5f", dx, dy, dz))
            end
            att = new_child(field, "Attribute")
            set_attributes(att; AttributeType="Scalar", Center="Node", Name=var)
            slab = new_child(att, "DataItem")
            set_attributes(slab, ItemType="HyperSlab", Dimensions=@sprintf("%i %i %i", fielddims["z"], fielddims["y"], fielddims["x"]), Type="HyperSlab")
            slabstep = new_child(slab, "DataItem")
            set_attributes(slabstep, Dimensions="3 3")
            add_text(slabstep, @sprintf("0 0 0 1 1 1 %i %i %i", fielddims["z"], fielddims["y"], fielddims["x"]))
            data = new_child(slab, "DataItem")
            set_attributes(data; Dimensions=@sprintf("%i %i %i", fielddims["z"], fielddims["y"], fielddims["x"]), Format="HDF")
            add_text(data, "fields/"*@sprintf("%05i", i)*"_"*var*".h5:/var")
        end
        save_file(xdoc, folder*"/outputdir/"*var*"_fields.xmf")
    end
end

end # module

File: tools/AFiDTools.py
# Tools for pre- and post-processing AFiD data

import h5py
import numpy as np
import os

class Grid:
    def __init__(self, xm, xmr, xc, xcr, ym, ymr, yc, ycr, zm, zmr, zc, zcr):
        self.xm = xm
        self.xmr= xmr
        self.xc = xc
        self.xcr= xcr
        self.ym = ym
        self.ymr= ymr
        self.yc = yc
        self.ycr= ycr
        self.zm = zm
        self.zmr= zmr
        self.zc = zc
        self.zcr= zcr

def read_grid(folder):
    with h5py.File(folder+"/outputdir/cordin_info.h5","r") as f:
        xm = f["xm"][()]
        xc = f["xc"][()]
        ym = f["ym"][()]
        zm = f["zm"][()]
        if "xmr" in list(f.keys()):
            xmr = f["xmr"][()]
            xcr = f["xcr"][()]
            ymr = f["ymr"][()]
            zmr = f["zmr"][()]
        else:
            xmr, xcr = np.array([]), np.array([])
            ymr, zmr = np.array([]), np.array([])
    if ym.size > 1:
        dy = ym[1] - ym[0]
        yc = np.arange(0, dy*(ym.size+1), dy)
    else: yc = []
    if ymr.size > 1:
        dyr = ymr[1] - ymr[0]
        ycr = np.arange(0, dyr*(ymr.size+1), dyr)
    else: ycr = []
    if zm.size > 1:
        dz = zm[1] - zm[0]
        zc = np.arange(0, dz*(zm.size+1), dz)
    else: zc = []
    if zmr.size > 1:
        dzr = zmr[1] - zmr[0]
        zcr = np.arange(0, dzr*(zmr.size+1), dzr)
    else: zcr = []
    return Grid(xm, xmr, xc, xcr, ym, ymr, yc, ycr, zm, zmr, zc, zcr)

def read_mean(folder, varname):
    with h5py.File(folder+"/outputdir/means.h5","r") as f:
        samplist = list(f[varname].keys())
        Nsamp = len(samplist)
        nx = f[varname+"/"+samplist[0]].size
        var = np.zeros((nx,Nsamp))
        i = 0
        for num in samplist:
            var[:,i] = f[varname+"/"+num][()]
            i = i + 1
    return var

def read_cut(folder, var, idx, plane):
    varname = var+"/"+"%05d" % idx
    with h5py.File(folder+"/outputdir/flowmov/movie_"+plane+"cut.h5","r") as f:
        A = np.array(f[varname][()])
    return A

def continua_master_from_input(folder):
    with open(folder+"/bou.in", "r") as f:
        for i, line in enumerate(f):
            if i==2:
                nxm, nym, nzm = [float(n) for n in line.split()[0:3]]
            if i==6:
                nxmr, nymr, nzmr = [float(n) for n in line.split()[1:4]]
            if i==26:
                ylen, zlen = [float(n) for n in line.split()[1:3]]
            if i==30:
                istr3, str3, istr3r = [float(n) for n in line.split()]
        time = 0.0
        with h5py.File(folder+"/outputdir/continua_master.h5","w") as f:
            f["nx"], f["ny"], f["nz"] = nxm + 1, nym + 1, nzm + 1
            f["nxr"], f["nyr"], f["nzr"] = nxmr + 1, nymr + 1, nzmr + 1
            f["ylen"], f["zlen"] = ylen, zlen
            f["istr3"], f["str3"], f["istr3r"] = istr3, str3, istr3r
            f["time"] = time
    return

def write_continua(folder, varname, var):
    with h5py.File(folder+"/outputdir/continua_"+varname+".h5","w") as f:
        f["var"] = var
    return

def z_vorticity(folder, idx, grid):
    vx = read_cut(folder, "vx", idx, "z")
    vy = read_cut(folder, "vy", idx, "z")

    dvxdy, dvydx = np.zeros(vx.shape), np.zeros(vy.shape)
    dy = (grid.ym[1] - grid.ym[0])

    dvxdy[1:-1,:] = (vx[2:,:] - vx[:-2,:])/2/dy
    dvxdy[0,:] = (vx[1,:] - vx[-1,:])/2/dy
    dvxdy[-1,:] = (vx[0,:] - vx[-2,:])/2/dy

    dvydx[:, 1:-1] = (vy[:,2:] - vy[:,:-2])/(grid.xm[2:] - grid.xm[:-2])
    dvydx[:, 0] = (vy[:,1] - 0)/(grid.xm[1] - 0)
    dvydx[:, -1] = (0 - vy[:,-2])/(1 - grid.xm[-2])

    dyu = 0.5*(dvxdy[:,1:] + dvxdy[:,:-1])

    dxv = np.zeros(vy.shape)
    dxv[:-1,:] = 0.5*(dvydx[1:,:] + dvydx[:-1,:])
    dxv[-1,:] = 0.5*(dvydx[0,:] + dvydx[-1,:])

    ζ = dxv - dyu

    return ζ

from xml.etree.ElementTree import Element, SubElement, Comment
from xml.dom import minidom
from xml.etree import ElementTree

def generate_cut_xmf(folder, plane):
    """
    Generates an xmf file in the Xdmf format to allow reading of
    the movie slices in ParaView. Specify the `plane` ("x", "y", "z")
    of the slices and the `folder` containing the simulation.
    """

    # Read the grid data from the simulation
    grid = read_grid(folder)
    nxm, nym, nzm = grid.xm.size, grid.ym.size, grid.zm.size
    nxmr, nymr, nzmr = grid.xmr.size, grid.ymr.size, grid.zmr.size

    # Store the dimensions of the slices in dims and dimsr
    if plane=="x":
        dims, dimsr = (nzm, nym), (nzmr, nymr)
    elif plane=="y":
        dims, dimsr = (nzm, nxm), (nzmr, nxmr)
    elif plane=="z":
        dims, dimsr = (nym, nxm), (nymr, nxmr)
    else:
        raise NameError(plane)

    # Count the number of movie samples recorded in the time series
    with h5py.File(folder+"/outputdir/flowmov/movie_"+plane+"cut.h5","r") as f:
        Nsamp = len(list(f["temp"].keys()))
    varlist = ["vx", "vy", "vz", "temp"]

    # Check the time interval used for frame writing from bou.in
    with open(folder+"/bou.in","r") as f:
        for i, line in enumerate(f):
            if i==22:
                t_frame = float(line.split()[1])


    ### BUILD THE XMF STRUCTURE ###
    Xdmf = Element("Xdmf")

    top_domain = SubElement(Xdmf, "Domain")

    # Create Grid element to store the time series
    time_series = SubElement(top_domain, "Grid", attrib={
        "Name":"Time", "GridType":"Collection", "CollectionType":"Temporal"
    })

    # Create empty tuples to build element tree with
    cuts, geom, xdata, ydata, zdata = (), (), (), (), ()
    # Create empty tuples for each variable
    var_att, var_data = [(), (), (), ()], [(), (), (), ()]
    for i in range(Nsamp):
        cuts = cuts + (SubElement(time_series, "Grid", attrib={
            "Name":plane+"cut", "GridType":"Uniform"
        }),)

        SubElement(cuts[i], "Time", attrib={"Value":"%.1f" % float(t_frame*i)})

        SubElement(cuts[i], "Topology", attrib={
            "TopologyType":"3DRectMesh", "Dimensions":"%i %i" % dims
        })
        
        geom = geom + (SubElement(cuts[i], "Geometry", attrib={
            "GeometryType":"VXVYVZ"
        }),)
        zdata = zdata + (SubElement(geom[i], "DataItem", attrib={"Dimensions":"1"}),)
        zdata[i].text = "0.0"
        ydata = ydata + (SubElement(geom[i], "DataItem", attrib={
            "Dimensions":"%i" % dims[0], "Format":"HDF"
        }),)
        if plane=="z":
            ydata[i].text = "../cordin_info.h5:/ym"
        else:
            ydata[i].text = "../cordin_info.h5:/zm"
        xdata = xdata + (SubElement(geom[i], "DataItem", attrib={
            "Dimensions":"%i" % dims[1], "Format":"HDF"
        }),)
        if plane=="x":
            xdata[i].text = "../cordin_info.h5:/ym"
        else:
            xdata[i].text = "../cordin_info.h5:/xm"
        
        for j, var in enumerate(varlist):
            var_att[j] = var_att[j] + (SubElement(cuts[i], "Attribute", attrib={
                "Name":var, "AttributeType":"Scalar", "Center":"Node"
            }),)
            var_data[j] = var_data[j] + (SubElement(var_att[j][i], "DataItem", attrib={
                "Dimensions":"%i %i" % dims, "Format":"HDF"
            }),)
            var_data[j][i].text = "movie_"+plane+"cut.h5:/"+var+"/%05i" % i

    # Convert xmf structure to string
    rough_xmf = ElementTree.tostring(Xdmf, "utf-8")

    # Format string for readability and write to file
    formatted_xmf = minidom.parseString(rough_xmf)
    with open(folder+"/outputdir/flowmov/movie_"+plane+"cut.xmf","w") as f:
        f.write(formatted_xmf.toprettyxml(indent="  "))



def generate_field_xmf(folder, var):
    """
    Generates an xmf file in the Xdmf format to allow reading of
    the 3D fields in ParaView. Specify the variable `var`
    ("vx", "vy", "vz", "temp", "sal", "phi") and the `folder`
    containing the simulation.
    """

    # Read the grid data from the simulation
    grid = read_grid(folder)
    nxm, nym, nzm = grid.xm.size, grid.ym.size, grid.zm.size
    nxmr, nymr, nzmr = grid.xmr.size, grid.ymr.size, grid.zmr.size

    # Store the appropriate grid sizes and names based on the variable
    fulldims = (nzm, nym, nxm+1)
    xx, yy, zz = "xm", "ym", "zm"
    if var=="vx":
        dims = (nzm, nym, nxm+1)
        xx = "xc"
    elif var in "phisal":
        dims = (nzmr, nymr, nxmr)
        fulldims = (nzmr, nymr, nxmr+1)
        xx, yy, zz = "xmr", "ymr", "zmr"
    else:
        dims = (nzm, nym, nxm)
        if var=="vy":
            yy = "yc"
        elif var=="vz":
            zz = "zc"
    
    # Collect indices of saved fields
    samplist = []
    for file in os.listdir(folder+"/outputdir/fields"):
        if var in file:
            samplist.append(int(file[:5]))

    # Check the time interval used for field writing from bou.in
    with open(folder+"/bou.in","r") as f:
        for i, line in enumerate(f):
            if i==22:
                t_field = float(line.split()[2])


    ### BUILD THE XMF STRUCTURE ###
    Xdmf = Element("Xdmf")

    top_domain = SubElement(Xdmf, "Domain")

    # Create Grid element to store the time series
    time_series = SubElement(top_domain, "Grid", attrib={
        "Name":"Time", "GridType":"Collection", "CollectionType":"Temporal"
    })

    fields, geom, xdata, ydata, zdata = (), (), (), (), ()
    var_att, var_slab, slab_data, var_data = (), (), (), ()

    for i in samplist:
        fields = fields + (SubElement(time_series, "Grid", attrib={
            "Name":"field", "GridType":"Uniform"
        }),)

        SubElement(fields[i], "Time", attrib={"Value":"%.1f" % float(t_field*i)})

        SubElement(fields[i], "Topology", attrib={
            "TopologyType":"3DRectMesh", "Dimensions":"%i %i %i" % dims
        })
        
        geom = geom + (SubElement(fields[i], "Geometry", attrib={
            "GeometryType":"VXVYVZ"
        }),)
        zdata = zdata + (SubElement(geom[i], "DataItem", attrib={
            "Dimensions":"%i" % dims[2], "Format":"HDF"
        }),)
        zdata[i].text = "cordin_info.h5:/"+xx
        ydata = ydata + (SubElement(geom[i], "DataItem", attrib={
            "Dimensions":"%i" % dims[1], "Format":"HDF"
        }),)
        ydata[i].text = "cordin_info.h5:/"+yy
        xdata = xdata + (SubElement(geom[i], "DataItem", attrib={
            "Dimensions":"%i" % dims[0], "Format":"HDF"
        }),)
        xdata[i].text = "cordin_info.h5:/"+zz
        
        var_att = var_att + (SubElement(fields[i], "Attribute", attrib={
            "Name":var, "AttributeType":"Scalar", "Center":"Node"
        }),)
        var_slab = var_slab + (SubElement(var_att[i], "DataItem", attrib={
            "ItemType":"HyperSlab", "Dimensions":"%i %i %i" % dims,
            "Type": "HyperSlab"
        }),)
        slab_data = slab_data + (SubElement(var_slab[i], "DataItem", attrib={
            "Dimensions":"3 3", "Format":"XML"
        }),)
        slab_data[i].text = "0 0 0 1 1 1 %i %i %i" % dims
        var_data = var_data + (SubElement(var_slab[i], "DataItem", attrib={
            "Dimensions":"%i %i %i" % fulldims, "Format":"HDF"
        }),)
        var_data[i].text = "fields/%05i_" % i + var +".h5:/var"

    # Convert xmf structure to string
    rough_xmf = ElementTree.tostring(Xdmf, "utf-8")

    # Format string for readability and write to file
    formatted_xmf = minidom.parseString(rough_xmf)
    with open(folder+"/outputdir/"+var+"_fields.xmf","w") as f:
        f.write(formatted_xmf.toprettyxml(indent="  "))
